Content.Shared/NameModifier/EntitySystems/NameModifierSystem.cs

```diff
@@ -0,0 +1,131 @@
+using System.Linq;
+using Content.Shared.Inventory;
+using Content.Shared.NameModifier.Components;
+using Robust.Shared.Collections;
+
+namespace Content.Shared.NameModifier.EntitySystems;
+
+public sealed partial class NameModifierSystem : EntitySystem
+{
+    [Dependency] private readonly MetaDataSystem _metaData = default!;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+
+        SubscribeLocalEvent<NameModifierComponent, EntityRenamedEvent>(OnEntityRenamed);
+    }
+
+    private void OnEntityRenamed(Entity<NameModifierComponent> entity, ref EntityRenamedEvent args)
+    {
+        SetBaseName((entity, entity.Comp), args.NewName);
+        RefreshNameModifiers((entity, entity.Comp));
+    }
+
+    private void SetBaseName(Entity<NameModifierComponent> entity, string name)
+    {
+        if (name == entity.Comp.BaseName)
+            return;
+
+        // Set the base name to the new name
+        entity.Comp.BaseName = name;
+        Dirty(entity);
+    }
+
+    /// <summary>
+    /// Raises a <see cref="RefreshNameModifiersEvent"/> to gather modifiers and
+    /// updates the entity's name to its base name with modifiers applied.
+    /// This will add a <see cref="NameModifierComponent"/> if any modifiers are added.
+    /// </summary>
+    /// <remarks>
+    /// Call this to update the entity's name when adding or removing a modifier.
+    /// </remarks>
+    public void RefreshNameModifiers(Entity<NameModifierComponent?> entity)
+    {
+        Resolve(entity, ref entity.Comp, logMissing: false);
```

Comment:
is this resolve here doing anything when you just add the component if it doesnt have it anyways? (the xmldoc of this method could be slightly clear about this fact too, it implies it but doesnt say it directly)

Reply:
Surprisingly, it does seem to be necessary. Without it, another system calling this method must check for the component and pass it in if it exists, which is kinda terrible.

If they don't, this code will see ent.Comp as null, even though the entity does have the component. That causes all kinds of problems, with BaseName getting set to include modifiers, causing them to get permanently added to the name, and to stack up with repeated refreshes - in testing, I had a glued glued glued glued crowbar that wasn't actually glued.

The alternative would be to use EnsureComp at the top of the method and then RemComp later if the component isn't needed, but I assume that adding and removing a new component is less efficient than calling Resolve.

Reply:
I mean i guess whats weird is that you arent using the resolve return value so you are basically just doing a TryComp without just directly calling it, so i would either do that or make use of the resolve to the fullest and do something like :
```c#
        var meta = MetaData(entity);
        var baseName = meta.EntityName;
        if (Resolve(entity, ref entity.Comp, logMissing: false))
            baseName = entity.Comp.BaseName;
```